### PR TITLE
Annotate System.ComponentModel.Composition for nullable

### PR DESCRIPTION
**Pull request by buyaa-n**
Contributes to https://github.com/dotnet/corefx/issues/40623

### CODE REVIEW

File: src/libraries/System.ComponentModel.Composition/System.ComponentModel.Composition.sln [hunk 1]
```diff
@@ -39,8 +39,8 @@ Global
 		{396D6EBF-60BD-4DAF-8783-FB403E070A56}.Debug|Any CPU.Build.0 = netstandard2.0-Debug|Any CPU
 		{396D6EBF-60BD-4DAF-8783-FB403E070A56}.Release|Any CPU.ActiveCfg = netstandard2.0-Release|Any CPU
 		{396D6EBF-60BD-4DAF-8783-FB403E070A56}.Release|Any CPU.Build.0 = netstandard2.0-Release|Any CPU
-		{2D694AC8-A12F-4622-9405-74E20EC40C3A}.Debug|Any CPU.ActiveCfg = netcoreapp5.0-Debug|Any CPU
-		{2D694AC8-A12F-4622-9405-74E20EC40C3A}.Debug|Any CPU.Build.0 = netcoreapp5.0-Debug|Any CPU
+		{2D694AC8-A12F-4622-9405-74E20EC40C3A}.Debug|Any CPU.ActiveCfg = netcoreapp2.0-Debug|Any CPU
```
Comment (safern) [inline src/libraries/System.ComponentModel.Composition/System.ComponentModel.Composition.sln]: It seems like this wasn't intentional, could you please revert?

Comment (buyaa-n) [reply src/libraries/System.ComponentModel.Composition/System.ComponentModel.Composition.sln]: whoops, yes it is not intended, will revert

File: src/libraries/System.ComponentModel.Composition/src/Microsoft/Internal/Collections/CollectionServices.cs [hunk 2]
```diff
@@ -269,7 +260,7 @@ public static bool IsArrayEqual<T>(this T[] thisArray, T[] thatArray)
 
             for (int i = 0; i < thisArray.Length; i++)
             {
-                if (!thisArray[i].Equals(thatArray[i]))
+                if (!thisArray[i]!.Equals(thatArray[i]))
```
Comment (buyaa-n) [inline src/libraries/System.ComponentModel.Composition/src/Microsoft/Internal/Collections/CollectionServices.cs]: this is weird, didn't see warning in this case before

File: src/libraries/System.ComponentModel.Composition/src/System/ComponentModel/Composition/Hosting/CatalogExportProvider.cs [hunk 3]
```diff
@@ -159,7 +157,7 @@ public ExportProvider SourceProvider
                 ThrowIfDisposed();
                 using (_lock.LockStateForRead())
                 {
-                    return _sourceProvider;
+                    return _sourceProvider!;  // Should be [DisallowNull]
```
Comment (buyaa-n) [inline src/libraries/System.ComponentModel.Composition/src/System/ComponentModel/Composition/Hosting/CatalogExportProvider.cs]: I think this property should be annotated as nullable with `[DisallowNull]` as setter throw on null but getter could return null before it set with a non null value. 

The problem here with `[DisallowNull]` attribute setter `value` not became non null, and warning

File: src/libraries/System.ComponentModel.Composition/src/System/ComponentModel/Composition/Hosting/CompositionServices.cs [hunk 4]
```diff
@@ -307,10 +307,10 @@ internal static void TryExportMetadataForMember(this MemberInfo member, out IDic
 
                             if (reservedMetadataNames.Contains(pi.Name, StringComparers.MetadataKeyNames))
                             {
-                                throw ExceptionBuilder.CreateDiscoveryException(SR.Discovery_ReservedMetadataNameUsed, member.GetDisplayName(), provider.Name);
```
Comment (buyaa-n) [inline src/libraries/System.ComponentModel.Composition/src/System/ComponentModel/Composition/Hosting/CompositionServices.cs]: This is the case where `provider == null` seems this exception handling copied from other case but forgotten to change variable names accordingly, please review @safern @stephentoub

File: src/libraries/System.ComponentModel.Composition/src/System/ComponentModel/Composition/MetadataViewProvider.cs [hunk 5]
```diff
@@ -48,13 +49,13 @@ public static TMetadataView GetMetadataView<TMetadataView>(IDictionary<string, o
...
                     {
                         var implementationAttribute = metadataViewType.GetFirstAttribute<MetadataViewImplementationAttribute>();
+                        Debug.Assert(implementationAttribute != null);
                         proxyType = implementationAttribute.ImplementationType;
                         if (proxyType == null)
                         {
                             throw new CompositionContractMismatchException(SR.Format(
                                 SR.ContractMismatch_MetadataViewImplementationCanNotBeNull,
-                                metadataViewType.FullName,
-                                proxyType.FullName));
```
Comment (buyaa-n) [inline src/libraries/System.ComponentModel.Composition/src/System/ComponentModel/Composition/MetadataViewProvider.cs]: As per above condition `proxyType == null`, it is safe to remove it as the formatting string only expecting one parameter

File: src/libraries/System.ComponentModel.Composition/src/System/ComponentModel/Composition/ReflectionModel/ReflectionComposablePart.cs [hunk 6]
```diff
@@ -300,7 +301,7 @@ private object GetInstanceActivatingIfNeeded()
                 }
 
                 // if the instance has been already set
-                if (createdInstance == null)
```
Comment (buyaa-n) [inline src/libraries/System.ComponentModel.Composition/src/System/ComponentModel/Composition/ReflectionModel/ReflectionComposablePart.cs]: I think this is bug, if `createdInstance` is null, no need to release it, CC @safern @maryamariyan

File: src/libraries/System.ComponentModel.Composition/ref/System.ComponentModel.Composition.cs [hunk 7]
```diff
@@ -507,15 +511,14 @@ public abstract partial class ComposablePart
...
+        public virtual System.Collections.Generic.IDictionary<string, object?> Metadata { get { throw null; } }
         public virtual void Activate() { }
-        public abstract object GetExportedValue(System.ComponentModel.Composition.Primitives.ExportDefinition definition);
+        public abstract object? GetExportedValue(System.ComponentModel.Composition.Primitives.ExportDefinition definition);
         public abstract void SetImport(System.ComponentModel.Composition.Primitives.ImportDefinition definition, System.Collections.Generic.IEnumerable<System.ComponentModel.Composition.Primitives.Export> exports);
     }
     public abstract partial class ComposablePartCatalog : System.Collections.Generic.IEnumerable<System.ComponentModel.Composition.Primitives.ComposablePartDefinition>, System.Collections.IEnumerable, System.IDisposable
     {
         protected ComposablePartCatalog() { }
-        [System.ComponentModel.EditorBrowsableAttribute(System.ComponentModel.EditorBrowsableState.Never)]
```
Comment (eiriktsarpalis) [inline src/libraries/System.ComponentModel.Composition/ref/System.ComponentModel.Composition.cs]: This has been removed by the tooling and should probably be reinstated.

Comment (buyaa-n) [reply src/libraries/System.ComponentModel.Composition/ref/System.ComponentModel.Composition.cs]: Sure, thanks

File: src/libraries/System.ComponentModel.Composition/System.ComponentModel.Composition.sln [hunk 8]
```diff
@@ -39,8 +39,8 @@ Global
 		{396D6EBF-60BD-4DAF-8783-FB403E070A56}.Debug|Any CPU.Build.0 = netstandard2.0-Debug|Any CPU
 		{396D6EBF-60BD-4DAF-8783-FB403E070A56}.Release|Any CPU.ActiveCfg = netstandard2.0-Release|Any CPU
 		{396D6EBF-60BD-4DAF-8783-FB403E070A56}.Release|Any CPU.Build.0 = netstandard2.0-Release|Any CPU
-		{2D694AC8-A12F-4622-9405-74E20EC40C3A}.Debug|Any CPU.ActiveCfg = netcoreapp2.0-Debug|Any CPU
-		{2D694AC8-A12F-4622-9405-74E20EC40C3A}.Debug|Any CPU.Build.0 = netcoreapp2.0-Debug|Any CPU
+		{2D694AC8-A12F-4622-9405-74E20EC40C3A}.Debug|Any CPU.ActiveCfg = netcoreapp5.0-Debug|Any CPU
```
Comment (eiriktsarpalis) [inline src/libraries/System.ComponentModel.Composition/System.ComponentModel.Composition.sln]: Are these changes intentional?

File: src/libraries/System.ComponentModel.Composition/System.ComponentModel.Composition.sln [hunk 9]
```diff
@@ -1,6 +1,6 @@
 Microsoft Visual Studio Solution File, Format Version 12.00
-# Visual Studio Version 16
-VisualStudioVersion = 16.0.29630.52
+# Visual Studio Version 15
```
Comment (eiriktsarpalis) [inline src/libraries/System.ComponentModel.Composition/System.ComponentModel.Composition.sln]: Intentional?

File: src/libraries/System.ComponentModel.Composition/System.ComponentModel.Composition.sln (hunk 8 above)
Comment (buyaa-n) [reply src/libraries/System.ComponentModel.Composition/System.ComponentModel.Composition.sln]: nope, reverted

File: src/libraries/System.ComponentModel.Composition/System.ComponentModel.Composition.sln (hunk 9 above)
Comment (buyaa-n) [reply src/libraries/System.ComponentModel.Composition/System.ComponentModel.Composition.sln]: nope, reverted

File: src/libraries/Common/src/System/Composition/Diagnostics/DebuggerTraceWriter.cs [hunk 10]
```diff
@@ -10,7 +10,7 @@ namespace System.Composition.Diagnostics
 {
     internal sealed class DebuggerTraceWriter : TraceWriter
     {
-        private static readonly string s_sourceName = typeof(DebuggerTraceWriter).Assembly.GetName().Name;
+        private static readonly string s_sourceName = typeof(DebuggerTraceWriter).Assembly.GetName().Name!;
```
Comment (stephentoub) [inline src/libraries/Common/src/System/Composition/Diagnostics/DebuggerTraceWriter.cs]: This can just be a constant:
```C#
private const string SourceName = "System.ComponentModel.Composition";
```

File: src/libraries/System.ComponentModel.Composition/ref/System.ComponentModel.Composition.cs [hunk 11]
```diff
@@ -91,11 +91,11 @@ public enum CreationPolicy
     public partial class ExportAttribute : System.Attribute
     {
         public ExportAttribute() { }
-        public ExportAttribute(string contractName) { }
-        public ExportAttribute(string contractName, System.Type contractType) { }
-        public ExportAttribute(System.Type contractType) { }
-        public string ContractName { get { throw null; } }
-        public System.Type ContractType { get { throw null; } }
+        public ExportAttribute(string? contractName) { }
+        public ExportAttribute(string? contractName, System.Type? contractType) { }
```
Comment (stephentoub) [inline src/libraries/System.ComponentModel.Composition/ref/System.ComponentModel.Composition.cs]: Even if the implementation doesn't currently throw an exception for these being null, is the intention that they be non-null?  Seems like it would be a bug if null were passed.  Can you find any such usage?

File: src/libraries/System.ComponentModel.Composition/ref/System.ComponentModel.Composition.cs [hunk 12]
```diff
@@ -129,22 +129,22 @@ public partial interface ICompositionService
     public partial class ImportAttribute : System.Attribute
     {
         public ImportAttribute() { }
-        public ImportAttribute(string contractName) { }
-        public ImportAttribute(string contractName, System.Type contractType) { }
-        public ImportAttribute(System.Type contractType) { }
+        public ImportAttribute(string? contractName) { }
+        public ImportAttribute(string? contractName, System.Type? contractType) { }
```
Comment (stephentoub) [inline src/libraries/System.ComponentModel.Composition/ref/System.ComponentModel.Composition.cs]: Same question on these

File: src/libraries/System.ComponentModel.Composition/ref/System.ComponentModel.Composition.cs [hunk 13]
```diff
@@ -155,12 +155,12 @@ public partial class ImportingConstructorAttribute : System.Attribute
     public partial class ImportManyAttribute : System.Attribute
     {
         public ImportManyAttribute() { }
-        public ImportManyAttribute(string contractName) { }
-        public ImportManyAttribute(string contractName, System.Type contractType) { }
-        public ImportManyAttribute(System.Type contractType) { }
+        public ImportManyAttribute(string? contractName) { }
+        public ImportManyAttribute(string? contractName, System.Type? contractType) { }
```
Comment (stephentoub) [inline src/libraries/System.ComponentModel.Composition/ref/System.ComponentModel.Composition.cs]: Ditto

File: src/libraries/System.ComponentModel.Composition/ref/System.ComponentModel.Composition.cs [hunk 14]
```diff
@@ -174,9 +174,9 @@ public enum ImportSource
     public partial class InheritedExportAttribute : System.ComponentModel.Composition.ExportAttribute
     {
         public InheritedExportAttribute() { }
-        public InheritedExportAttribute(string contractName) { }
-        public InheritedExportAttribute(string contractName, System.Type contractType) { }
-        public InheritedExportAttribute(System.Type contractType) { }
+        public InheritedExportAttribute(string? contractName) { }
+        public InheritedExportAttribute(string? contractName, System.Type? contractType) { }
```
Comment (stephentoub) [inline src/libraries/System.ComponentModel.Composition/ref/System.ComponentModel.Composition.cs]: Ditto.  On all such cases.  I'll stop commenting on them.

File: src/libraries/System.ComponentModel.Composition/src/Microsoft/Internal/GenerationServices.cs [hunk 15]
```diff
@@ -15,7 +15,7 @@ namespace Microsoft.Internal
     internal static class GenerationServices
     {
         // Type.GetTypeFromHandle
-        private static readonly MethodInfo _typeGetTypeFromHandleMethod = typeof(Type).GetMethod("GetTypeFromHandle");
+        private static readonly MethodInfo _typeGetTypeFromHandleMethod = typeof(Type).GetMethod("GetTypeFromHandle")!;
```
Comment (stephentoub) [inline src/libraries/System.ComponentModel.Composition/src/Microsoft/Internal/GenerationServices.cs]: Not necessarily for this PR, but at some point it'd be nice to fix the naming here, e.g. s_typeGetTypeFromHandleMethod

File: src/libraries/System.ComponentModel.Composition/ref/System.ComponentModel.Composition.cs (hunk 11 above)
Comment (buyaa-n) [reply src/libraries/System.ComponentModel.Composition/ref/System.ComponentModel.Composition.cs]: Didn't found real usage with null, only found unit test with null 😄. 
But the doc says use null or empty string if want to use default, isn't that mean intention was allow null? https://github.com/dotnet/runtime/blob/510113bc7f0a4a505ab0ccb90b0a593c3943baae/src/libraries/System.ComponentModel.Composition/src/System/ComponentModel/Composition/ExportAttribute.cs#L99

Comment (stephentoub) [reply src/libraries/System.ComponentModel.Composition/ref/System.ComponentModel.Composition.cs]: > isn't that mean intention was allow null?

Yup, ok, thanks.

File: src/libraries/System.ComponentModel.Composition/ref/System.ComponentModel.Composition.cs (hunk 12 above)
Comment (buyaa-n) [reply src/libraries/System.ComponentModel.Composition/ref/System.ComponentModel.Composition.cs]: This one also only unit test, this method doesn't have spceific doc, but above method documented to allow null
https://github.com/dotnet/runtime/blob/510113bc7f0a4a505ab0ccb90b0a593c3943baae/src/libraries/System.ComponentModel.Composition/src/System/ComponentModel/Composition/ImportAttribute.cs#L73

File: src/libraries/System.ComponentModel.Composition/ref/System.ComponentModel.Composition.cs (hunk 14 above)
Comment (buyaa-n) [reply src/libraries/System.ComponentModel.Composition/ref/System.ComponentModel.Composition.cs]: This one has documented to accept nulls

File: src/libraries/System.ComponentModel.Composition/ref/System.ComponentModel.Composition.cs (hunk 13 above)
Comment (buyaa-n) [reply src/libraries/System.ComponentModel.Composition/ref/System.ComponentModel.Composition.cs]: This one also have doc, but above and below methods documented to allow null

File: src/libraries/System.ComponentModel.Composition/src/Microsoft/Internal/GenerationServices.cs (hunk 15 above)
Comment (buyaa-n) [reply src/libraries/System.ComponentModel.Composition/src/Microsoft/Internal/GenerationServices.cs]: I can do that here

**Review by eiriktsarpalis — APPROVED** [review]

File: src/libraries/Common/src/System/Composition/Diagnostics/DebuggerTraceWriter.cs [hunk 16]
```diff
@@ -10,7 +10,7 @@ namespace System.Composition.Diagnostics
 {
     internal sealed class DebuggerTraceWriter : TraceWriter
     {
-        private static readonly string s_sourceName = typeof(DebuggerTraceWriter).Assembly.GetName().Name;
+        private const string s_sourceName = "System.ComponentModel.Composition";
```
Comment (safern) [inline src/libraries/Common/src/System/Composition/Diagnostics/DebuggerTraceWriter.cs]: This file is in the `Common` path, I don't think this value will always be `System.ComponentModel.Composition`.

https://github.com/dotnet/runtime/search?q=%22DebuggerTraceWriter.cs+%22&type=

File: src/libraries/System.ComponentModel.Composition/src/Microsoft/Internal/AttributeServices.cs [hunk 17]
```diff
@@ -4,6 +4,7 @@
 
 using System.Linq;
 using System.Reflection;
+using System.Diagnostics.CodeAnalysis;
```
Comment (safern) [inline src/libraries/System.ComponentModel.Composition/src/Microsoft/Internal/AttributeServices.cs]: Why is this needed? I don't see any nullability attribute being used in this file.

File: src/libraries/System.ComponentModel.Composition/src/Microsoft/Internal/ReflectionServices.cs [hunk 18]
```diff
@@ -133,7 +137,8 @@ internal static IEnumerable<MethodInfo> GetAllMethods(this Type type)
         {
             IEnumerable<MethodInfo> declaredMethods = type.GetDeclaredMethods();
 
-            Type baseType = type.BaseType;
+            Type? baseType = type.BaseType;
+            Debug.Assert(baseType != null);
```
Comment (safern) [inline src/libraries/System.ComponentModel.Composition/src/Microsoft/Internal/ReflectionServices.cs]: Why not declare it non-nullable and bang the `type.BaseType` call given it is only used one time after this.

Comment (safern) [reply src/libraries/System.ComponentModel.Composition/src/Microsoft/Internal/ReflectionServices.cs]: Actually I just noticed it is used twice. Disregard 😄

File: src/libraries/Common/src/System/Composition/Diagnostics/DebuggerTraceWriter.cs (hunk 16 above)
Comment (buyaa-n) [reply src/libraries/Common/src/System/Composition/Diagnostics/DebuggerTraceWriter.cs]: Sounds better to revert back?  CC @stephentoub

File: src/libraries/System.ComponentModel.Composition/src/Microsoft/Internal/AttributeServices.cs (hunk 17 above)
Comment (buyaa-n) [reply src/libraries/System.ComponentModel.Composition/src/Microsoft/Internal/AttributeServices.cs]: I guess I was using attribute at first

File: src/libraries/Common/src/System/Composition/Diagnostics/DebuggerTraceWriter.cs (hunk 16 above)
Comment (stephentoub) [reply src/libraries/Common/src/System/Composition/Diagnostics/DebuggerTraceWriter.cs]: Hmm.  Ok.

**Review by safern — APPROVED** [review]
